subprojects/model-core/src/main/java/org/gradle/internal/snapshot/impl/SerializedValueSnapshot.java

```diff
@@ -73,6 +74,12 @@ private boolean hasSameSerializedValue(Object value, ValueSnapshot snapshot) {
             // Deserialize the old value and use the equals() implementation. This will be removed at some point
             Object oldValue = populateClass(value.getClass());
             if (oldValue.equals(value)) {
+                DeprecationLogger.deprecateIndirectUsage("Using objects as inputs which have a different serialized form but are equals")
+                    .withContext("Type '" + value.getClass().getName() + "' has a custom equals implementation.")
+                    .withAdvice("Use @Nested instead.")
```

Comment:
I think we should give a little bit more explanation here about how using `@Nested` solve this. How about:

```suggestion
                    .withAdvice("Declare the property as @Nested instead to expose its properties as inputs.")
```